Para Revisão campuscode/rails-guides-pt-BR#195 (#284)
campuscode/rails-guides-pt-BR: pt-BR/action_controller_overview.md

Para enviar um *hash*, você inclui o nome da chave dentro dos colchetes:

> ```HTML

Reviewer: ```suggestion
```html
```

Acho que aqui como faz parte de outro tópico podemos alterar futuramente.

Rescue
------

> Muito provavelmente sua aplicação irá conter bugs ou enviar excessões que precisam ser tratadas. Por exemplo, se o usuário acessar um link que não possui uma fonte no banco de dados, o Active Record enviará `ActiveRecord::RecordNotFound` como excessão.

Reviewer 2: Aqui só precisamos ajustar apenas a palavra `exceção` :D

```suggestion
Muito provavelmente sua aplicação irá conter bugs ou enviar exceções que precisam ser tratadas. Por exemplo, se o usuário acessar um link que não possui uma fonte no banco de dados, o Active Record enviará `ActiveRecord::RecordNotFound` como exceção.
```

> A exceção padrão do Rails apresenta a mensagem "500 Server Error" para todas as exceções. Se a requisição for feita localmente, um belo *traceback* e outras informações serão mostradas assim você pode verificar o que deu errado e tratar o problema. Se a requisição for remota o Rails apenas apresentará a mensagem "500 Server Error" para o usuário, ou um "404 Not Found" se houver um erro na rota ou o registro não puder ser encontrado. As vezes você pode querer customizar como esses erros são encontrados e como são apresentados ao usuário. Há diversos níveis de tratamento de excessões disponiveis em uma aplicação Rails:

Reviewer: @akaninja aqui acontece a mesma ocorrência do outro PR. O que acha? As frases entre aspas ficam em itálico também?

Reviewer 2: Eu acho positivo deixar em itálico porque ajuda a destacar que é uma expressão externa ao texto. Nesse caso em específico, não traduzimos as respostas padrão HTTP por ser um termo técnico.

Reviewer 3: Fiquei em dúvida aqui.... 

Minha primeira reação seria italizar o termo e deixar entre aspas. Mas quando usamos as aspas, já damos a entender que é uma expressão estrangeira, então ficaria excessivo termos as duas estilizações. 
Talvez a gente tenha que avaliar caso a caso e não impor uma regra... 
Nesse caso específico, acho que poderia ficar só entre aspas ou italizar e tirar as aspas. Tendo a preferir deixar só aspas, porque nesse caso não é um termo técnico, é mais uma citação da mensagem... 

Ajudei? Ou só atrapalhei?

> A excessão padrão do Rails apresenta a mensagem "500 Server Error" para todas as excessões. Se a requisição for feita localmente, um belo TRACEBACK e outras informações serão mostradas assim você pode verificar o que deu errado e tratar o problema. Se a requisição for remota o Rails apenas apresentará a mensagem "500 Server Error" para o usuário, ou um "404 Not Found" se houver um erro na rota ou o registro não puder ser encontrado. As vezes você pode querer customizar como esses erros são encontrados e como são apresentados ao usuário. Há diversos níveis de tratamento de excessões disponiveis em uma aplicação Rails:

Reviewer 2: Aqui precisamos ajustar a ocorrência da palavra `excessão -> exceção` 😄
Podemos também deixar o `traceback` em itálico e associar um artigo pra ele. @campuscode/guia-rails-team o que vocês acham? Eu normalmente falo `o traceback`, queria a opinião de vocês nisso.

```suggestion
A exceção padrão do Rails apresenta a mensagem "500 Server Error" para todas as exceções. Se a requisição for feita localmente, um belo TRACEBACK e outras informações serão mostradas assim você pode verificar o que deu errado e tratar o problema. Se a requisição for remota o Rails apenas apresentará a mensagem "500 Server Error" para o usuário, ou um "404 Not Found" se houver um erro na rota ou o registro não puder ser encontrado. As vezes você pode querer customizar como esses erros são encontrados e como são apresentados ao usuário. Há diversos níveis de tratamento de exceções disponiveis em uma aplicação Rails:
```

> ### Os templates 404 e 500 padrão

Reviewer: ```suggestion
### Os *Templates* 404 e 500 Padrão
```

Alguns estrangeirismos nós usamos itálico para marcação, acho que seria legal usar mesmo nos títulos. O que acha?

> Por padrão uma aplicação em produção irá renderizar uma mensagem em um template de erro 404 ou 500, no ambiente de produção todas as mensagens de erro são disparadas. Essas mensagens são armazenadas em templates estáticos de HTML na pasta *public*, em `404.html` e `500.html` respectivamente. Você pode customizar essas páginas e adcionar algumas estilizações, mas lembre-se elas são HTML estático; i.e. você não pode usar ERB, SCSS, CoffeeScript, ou layouts para elas.

Reviewer: ```suggestion
Por padrão uma aplicação em produção irá renderizar uma mensagem em um template de erro 404 ou 500, no ambiente de produção todas as mensagens de erro são disparadas. Essas mensagens são armazenadas em templates estáticos de HTML na pasta *public*, em `404.html` e `500.html` respectivamente. Você pode customizar essas páginas e adicionar algumas estilizações, mas lembre-se elas são HTML estático; i.e. você não pode usar ERB, SCSS, CoffeeScript, ou layouts para elas.
```

Aqui tem um pequeno typo, pode dar uma olhada?

> Por padrão uma aplicação em produção irá renderizar uma mensagem em um template de erro 404 ou 500, no ambiente de produção todas as mensagens de erro são disparadas. Essas mensagens são armazenadas em templates estáticos de html na pasta *public*, em `404.html` e `500.html` respectivamente. Você pode customizar essas páginas e adcionar algumas estilizações, mas lembre-se elas são html estático; i.e. você não pode usar ERB, SCSS, CoffeeScript, ou layouts para elas.

Reviewer 2: Podemos ajustar o HTML pra manter a mesma formatação do texto original: 💅 

```suggestion
Por padrão uma aplicação em produção irá renderizar uma mensagem em um template de erro 404 ou 500, no ambiente de produção todas as mensagens de erro são disparadas. Essas mensagens são armazenadas em templates estáticos de html na pasta *public*, em `404.html` e `500.html` respectivamente. Você pode customizar essas páginas e adcionar algumas estilizações, mas lembre-se elas são HTML estático; i.e. você não pode usar ERB, SCSS, CoffeeScript, ou layouts para elas.
```

### `rescue_from`

> Se você quiser fazer algo mais elaborado quando estiver lidando com erros, você pode usar `rescue_from`, que trata as exceções de um certo tipo (ou de vários tipos) em um controller inteiro e nas subclasses.

Reviewer: ```suggestion
Se você quiser fazer algo mais elaborado quando estiver lidando com erros, você pode usar `rescue_from`, que trata as exceções de um certo tipo (ou de vários tipos) em um *controller* inteiro e nas subclasses.
```

Costumamos usar controller em itálico então fiz um pequeno comentário só pra mantermos o padrão mas nada demais, pode dar uma olhada?

> Se você quiser fazer algo mais elaborado quando estiver lidando com erros, você pode usar `rescue_from`, qu trata as excessões de um certo tipo (ou de vários tipos) em um controller inteiro e nas subclasses.

Reviewer 2: Aqui é a mesma coisa do comentário anterior:

```suggestion
Se você quiser fazer algo mais elaborado quando estiver lidando com erros, você pode usar `rescue_from`, qu trata as exceções de um certo tipo (ou de vários tipos) em um controller inteiro e nas subclasses.
```

> Quando uma excessão acontece e é pega por uma diretiva `rescue_from`, o objeto da excessão é passado ao handler. O handler pode ser um método ou um objeto `Proc` passado com a opção `:with`. Você também pode usar um bloco diretamente ao invés de um objetco `Proc`.

Reviewer 2: Aqui é a mesma coisa do comentário anterior, junto com colocar `handler`em itálico e um pequeno typo que eu achei: 💅 

```suggestion
Quando uma exceção acontece e é pega por uma diretiva `rescue_from`, o objeto da exceção é passado ao *handler*. O *handler* pode ser um método ou um objeto `Proc` passado com a opção `:with`. Você também pode usar um bloco diretamente ao invés de um objeto `Proc`.
```

> Quando uma excessão acontece e é pega por uma diretiva `rescue_from`, o objeto da exceção é passado ao *handler*. O *handler* pode ser um método ou um objeto `Proc` passado com a opção `:with`. Você também pode usar um bloco diretamente ao invés de um objetco `Proc`.

Reviewer: ```suggestion
Quando uma exceção acontece e é pega por uma diretiva `rescue_from`, o objeto da exceção é passado ao *handler*. O *handler* pode ser um método ou um objeto `Proc` passado com a opção `:with`. Você também pode usar um bloco diretamente ao invés de um objeto `Proc`.
```

Acho que tem pequeno typo aqui, pode dar uma olhada?

end
```

> É claro, que este exemplo não é nada elaborado e não melhora muito a forma original de lidar com os erros, mas uma vez que você capture todas essas excessões você é livre para fazer o que quiser com elas. Por exemplo, você pode criar uma exceção personalizada para quando o usuário não tem acesso a uma parte da aplicação:

Reviewer: ```suggestion
É claro, que este exemplo não é nada elaborado e não melhora muito a forma original de lidar com os erros, mas uma vez que você capture todas essas exceções você é livre para fazer o que quiser com elas. Por exemplo, você pode criar uma exceção personalizada para quando o usuário não tem acesso a uma parte da aplicação:
```

Aqui parece ter um pequeno typo, pode dar uma olhada?

Reviewer 2: Boa @HenriqueMorato 😄 
Eu não tinha visto essa ocorrência ainda.

> É claro, que este exemplo não é nada elaborado e não melhora muito a forma original de lidar com os erros, mas uma vez que você capture todas essas excessões você é livre para fazer o que quiser com elas. Por exemplo, você pode criar uma excessão personalizada para quando o usuário não tem acesso a uma parte da aplicação:

Reviewer 2: ```suggestion
É claro, que este exemplo não é nada elaborado e não melhora muito a forma original de lidar com os erros, mas uma vez que você capture todas essas exceções você é livre para fazer o que quiser com elas. Por exemplo, você pode criar uma exceção personalizada para quando o usuário não tem acesso a uma parte da aplicação:
```

Author: vou fazer todas, creio que não prestei atenção o bastante nos detalhes, mas todas me parecem validas

Author: acho que fiz uns commits meio ruins, preciso aprender a remover eles na hora da entrega... Mas em geral todos os seus comentários foram corrigidos pelo visto e achei letras comidas no meio do caminho que eu corrigi

Reviewer 2: Então, acho que agora não precisa se preocupar com isso. :)
Na hora de fazer o merge do PR, o time faz squash dos commits. Isso significa que eles juntam todas as alterações num commit só, aí fica tudo dentro de uma unidade.

Se você quiser fazer essa organização você mesmo, dá pra fazer isso com o `git rebase`.

- Faça o pull do branch `upstream` no seu fork:

`$ git checkout master`
`$ git pull upstream`

- Depois passe pro seu branch:

`$ git checkout Action-Controller-Overview->-14.-Rescue`

- Execute:
`$ git rebase -i --autosquash master`

Pra mais informações, vc pode olhar [aqui](https://help.github.com/pt/github/using-git/using-git-rebase-on-the-command-line)

Reviewer 2: Você também pode marcar os próximos commits como `fixup`. Isso faz com que eles sejam mesclados de forma automática na hora de fazer rebase:

`$ git commit --fixup`

https://saulosantiago.js.org/git/git-fixup-com-autosquash/

end
```

> AVISO: Ao usar `rescue_from` com `Exception` ou `StandardError` pode causar efeitos colaterais já que previne o Rails de lidar com as excessões apropriadamente. Dessa forma, não é recomendado fazer sem uma boa razão.

Reviewer: ```suggestion
WARNING: Ao usar `rescue_from` com `Exception` ou `StandardError` pode causar efeitos colaterais já que previne o Rails de lidar com as exceções apropriadamente. Dessa forma, não é recomendado fazer sem uma boa razão.
```

Aqui tem um pequeno typo, pode dar uma olhada?

Essas marcações de WARNING, NOTE, etc nós não traduzimos porque elas geram marcações no HTML final, você pode ler mais sobre isso [aqui](https://github.com/campuscode/rails-guides-pt-BR/blob/master/TRANSLATION_MANUAL.md#tip-note-e-warning)

> NOTA: Quando rodando em ambiente de desenvolvimento, todos os erros

Reviewer: ```suggestion
NOTE: Quando rodando em ambiente de desenvolvimento, todos os erros
```

Reviewer: @marcospedro97 pode dar uma olhada nesse?

`ActiveRecord::RecordNotFound` renderizam uma página 404. A não ser que você precise de uma forma especifica de tratar isso você não precisa tratar isso.

> NOTA: Certas excessões são tratadas apenas pela classe `ApplicationController`, já que são acionadas antes do controller ser iniciado a exceção é executada.

Reviewer: ```suggestion
NOTE: Certas exceções são tratadas apenas pela classe `ApplicationController`, já que são acionadas antes do controller ser iniciado a exceção é executada.
```
